### PR TITLE
useSubStep hook

### PR DESCRIPTION
**Pull request by Swor71**
### Details

Added `useSubStep` hook to manage navigation between sub steps

### Fixed Issues

$ https://github.com/Expensify/App/issues/30239
PROPOSAL: 

### Tests

- [ ] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/hooks/useSubStep.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import {useState, useRef, useCallback} from 'react';
+import PropTypes from 'prop-types';
+
+const propTypes = PropTypes.shape({
+    bodyContent: PropTypes.arrayOf(PropTypes.element).isRequired,
+    startFrom: PropTypes.number.isRequired,
+    onFinished: PropTypes.func.isRequired,
+});
```
Comment (fabioh8010) [inline src/hooks/useSubStep.js:8]:
```suggestion
const propTypes = {
    bodyContent: PropTypes.arrayOf(PropTypes.element).isRequired,
    startFrom: PropTypes.number.isRequired,
    onFinished: PropTypes.func.isRequired,
};
```

Shouldn't we do like this?

Comment (fabioh8010) [reply src/hooks/useSubStep.js:8]: Also, it would be nice to add some description to each prop

File: src/hooks/useSubStep.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * @typedef {Object} ReturnType
+ * @property {React.ReactElement} componentToRender
+ * @property {boolean} isEditing
+ * @property {Function} nextScreen
+ * @property {Function} moveTo
+ */
```
Comment (fabioh8010) [inline src/hooks/useSubStep.js:16]: Where does this comment belong?

File: src/hooks/useSubStep.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * A hook which manages navigation between substeps
+ *
+ * @export
+ * @param {[React.ReactElement]} bodyContent
+ * @param {number} startFrom
+ * @param {Function} onFinished
+ * @return {ReturnType}
+ */
+
```
Comment (fabioh8010) [inline src/hooks/useSubStep.js:27]:
```suggestion
```

File: src/hooks/useSubStep.js (hunk 2 above)
Comment (Swor71) [reply src/hooks/useSubStep.js]: not sure, if I understand the question, this typedef defines the return type for the hook so we get nice intellisense when using it. Would you like it to be in some other place?

Comment (fabioh8010) [reply src/hooks/useSubStep.js]: Oh I see now, it's connected to the other JSDOC comment, nice!

File: src/hooks/useSubStep.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import {useState, useRef, useCallback} from 'react';
+import PropTypes from 'prop-types';
+
+const propTypes = {
+    bodyContent: PropTypes.arrayOf(PropTypes.element).isRequired,
+    startFrom: PropTypes.number.isRequired,
+    onFinished: PropTypes.func.isRequired,
+};
```
Comment (shubham1206agra) [inline src/hooks/useSubStep.js:8]: Add comments about these props.

File: src/hooks/useSubStep.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * A hook which manages navigation between substeps
+ *
+ * @param {[React.ReactElement]} bodyContent an array of substep components
+ * @param {number} startFrom an index of the component from bodyContent array to start from
+ * @param {Function} onFinished a callback to be fired when pressing Confirm on the last substep screen
+ * @return {ReturnType}
+ */
+export default function useSubStep({bodyContent, startFrom = 0, onFinished}) {
```
Comment (shubham1206agra) [inline src/hooks/useSubStep.js:28]: Why does startFrom has a default value if the parameter is required?

File: src/hooks/useSubStep.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ * @property {Function} nextScreen a function to be called in order to navigate to the next substep screen
+ * @property {Function} moveTo a function to be called in order to navigate to a particular substep screen
+ */
+
+/**
+ * A hook which manages navigation between substeps
+ *
+ * @param {[React.ReactElement]} bodyContent an array of substep components
+ * @param {number} startFrom an index of the component from bodyContent array to start from
+ * @param {Function} onFinished a callback to be fired when pressing Confirm on the last substep screen
```
Comment (shubham1206agra) [inline src/hooks/useSubStep.js:25]: useSubStep has a single parameter as an object.
Can you correct this?

File: src/hooks/useSubStep.js (hunk 5 above)
Comment (Swor71) [reply src/hooks/useSubStep.js:28]: actually it should be optional, good catch, thanks

File: src/hooks/useSubStep.js (hunk 6 above)
Comment (Swor71) [reply src/hooks/useSubStep.js]: @shubham1206agra changed according to [jsdoc docs](https://jsdoc.app/tags-param.html#parameters-with-properties), is this better?

**Comment by Swor71** [comment]
@shubham1206agra please be aware that I've just pushed a change that introduces a function to go back to previous screen and `screenIndex` as these are required by the navigation flow.

**Comment by shubham1206agra** [comment]
I have one question here about the moveTo function.
When you use the moveTo function in between the process and then press next. Where should the screen go?
To the last step or the step where we previously were.

**Comment by Swor71** [comment]
> I have one question here about the moveTo function. When you use the moveTo function in between the process and then press next. Where should the screen go? To the last step or the step where we previously were.

@shubham1206agra 
`moveTo` will be used from the confirmation page in order to edit a specific sub step, it will then take the user back to the confirmation page (last step)

**Comment by shubham1206agra** [comment]
@Swor71 Do we need to put any check for this in `moveTo` function?
If not, I will approve the PR

**Comment by Swor71** [comment]
> @Swor71 Do we need to put any check for this in `moveTo` function? If not, I will approve the PR

the check for that is in the `nextScreen` function based on `isEditing` boolean so I think we should be good as is

**Comment by shubham1206agra** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by shubham1206agra — APPROVED** [review]

**Comment by shubham1206agra** [comment]
Since the error is unrelated, I am going to approve this request.
@mountiny Can you check the updated function and update the same in the docs, please?

File: src/hooks/useSubStep.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import {useState, useRef, useCallback} from 'react';
+import PropTypes from 'prop-types';
+
+const propTypes = {
+    /** an array of substep components */
```
Comment (mountiny) [inline src/hooks/useSubStep.js:5]: Could you fix this in next PR but the prop comments usually are capitalized

File: src/hooks/useSubStep.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    bodyContent: PropTypes.arrayOf(PropTypes.element).isRequired,
+
+    /** an index of the component from bodyContent array to start from */
```
Comment (mountiny) [inline src/hooks/useSubStep.js:8]: Same as above

**Review by mountiny — APPROVED** [review]
Thanks! just come comments for styles but please address that in another PR for the project, its minor

**Pull request merged**